File: Herradura_tests.py
'''
    Herradura KEx -- Security & Performance Tests (Python)

    Copyright (C) 2024-2026 Omar Alejandro Herrera Reyna

    This program is free software: you can redistribute it and/or modify
    it under the terms of the MIT License or the GNU General Public License
    as published by the Free Software Foundation, either version 3 of the License,
    or (at your option) any later version.

    Under the terms of the GNU General Public License, please also consider that:

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
'''

import os
import time

# ---------------------------------------------------------------------------
# BitArray class (self-contained, do not import from other files)
# ---------------------------------------------------------------------------

class BitArray:
    """Fixed-width bit string backed by a Python int.
    Supports XOR, in-place rotation, equality, and hex/bytes/uint I/O.
    Size must be a positive multiple of 8.
    """

    __slots__ = ('_val', '_size', '_mask')

    def __init__(self, size: int, value: int = 0):
        self._size = size
        self._mask = (1 << size) - 1
        self._val = int(value) & self._mask

    @property
    def uint(self) -> int:
        return self._val

    @uint.setter
    def uint(self, value: int):
        self._val = int(value) & self._mask

    @property
    def bytes(self) -> bytes:
        return self._val.to_bytes(self._size // 8, 'big')

    @bytes.setter
    def bytes(self, data: bytes):
        self._val = int.from_bytes(data, 'big') & self._mask

    @property
    def hex(self) -> str:
        return f'{self._val:0{self._size // 4}x}'

    def copy(self) -> 'BitArray':
        return BitArray(self._size, self._val)

    def rotated(self, n: int) -> 'BitArray':
        """Return a new BitArray rotated left by n bits (right if n < 0)."""
        n %= self._size
        if n == 0:
            return BitArray(self._size, self._val)
        return BitArray(self._size,
                        ((self._val << n) | (self._val >> (self._size - n))) & self._mask)

    def rol(self, n: int) -> None:
        """Rotate left in-place by n bits."""
        n %= self._size
        if n:
            self._val = ((self._val << n) | (self._val >> (self._size - n))) & self._mask

    def ror(self, n: int) -> None:
        """Rotate right in-place by n bits."""
        n %= self._size
        if n:
            self._val = ((self._val >> n) | (self._val << (self._size - n))) & self._mask

    def __xor__(self, other: 'BitArray') -> 'BitArray':
        return BitArray(self._size, self._val ^ other._val)

    def __ixor__(self, other: 'BitArray') -> 'BitArray':
        self._val ^= other._val
        return self

    def __eq__(self, other: object) -> bool:
        if isinstance(other, BitArray):
            return self._size == other._size and self._val == other._val
        return NotImplemented

    def __str__(self) -> str:
        return f'0x{self.hex}'

    def __repr__(self) -> str:
        return f'BitArray({self._size}, 0x{self.hex})'

    @classmethod
    def random(cls, size: int) -> 'BitArray':
        """Return a random BitArray of *size* bits using os.urandom."""
        ba = cls(size)
        ba.bytes = os.urandom(size // 8)
        return ba

    def flip_bit(self, pos: int) -> 'BitArray':
        """Return a new BitArray with bit *pos* (0=LSB) flipped."""
        return BitArray(self._size, self._val ^ (1 << pos))

    def popcount(self) -> int:
        """Return count of set bits."""
        return bin(self._val).count('1')


# ---------------------------------------------------------------------------
# FSCX functions (copy-based, self-contained)
# ---------------------------------------------------------------------------

def fscx(A: BitArray, B: BitArray) -> BitArray:
    """Full Surroundings Cyclic XOR: A ^ B ^ ROL(A) ^ ROL(B) ^ ROR(A) ^ ROR(B).
    Uses rotated() — does not mutate its inputs."""
    return A ^ B ^ A.rotated(1) ^ B.rotated(1) ^ A.rotated(-1) ^ B.rotated(-1)


def fscx_revolve(A: BitArray, B: BitArray, steps: int) -> BitArray:
    result = A.copy()
    for _ in range(steps):
        result = fscx(result, B)
    return result


def fscx_revolve_n(A: BitArray, B: BitArray, nonce: BitArray, steps: int) -> BitArray:
    result = A.copy()
    for _ in range(steps):
        result = fscx(result, B) ^ nonce
    return result


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def i_val(size: int) -> int:
    return size // 4


def r_val(size: int) -> int:
    return size * 3 // 4


SIZES = [64, 128, 256]

TARGET_SEC = 1.0


# ---------------------------------------------------------------------------
# Security tests
# ---------------------------------------------------------------------------

def test_noncommutativity():
    # FSCX(A,B) == FSCX(B,A) always (symmetric formula: A^B^ROL(A)^ROL(B)^ROR(A)^ROR(B)).
    # Asymmetry arises in FSCX_REVOLVE_N, where B is held constant across
    # iterations: FSCX_REVOLVE_N(A,B,N,n) != FSCX_REVOLVE_N(B,A,N,n) in general.
    # The nonce term T_n(N) cancels from both sides of the difference, so
    # commutativity is determined solely by the A and B inputs.
    print("[1] FSCX_REVOLVE_N non-commutativity: FSCX_REVOLVE_N(A,B,N,n) != FSCX_REVOLVE_N(B,A,N,n)")
    for size in SIZES:
        iv = i_val(size)
        comm = 0
        for _ in range(10000):
            a = BitArray.random(size)
            b = BitArray.random(size)
            n = BitArray.random(size)
            if fscx_revolve_n(a, b, n, iv) == fscx_revolve_n(b, a, n, iv):
                comm += 1
        status = "PASS" if comm == 0 else "FAIL"
        print(f"    bits={size:3d}  {comm:5d} / 10000 commutative  [{status}]")
    print()


def test_avalanche():
    # FSCX is a linear map over GF(2): output bit i depends only on input bits
    # i-1, i, i+1 (cyclic). Flipping one input bit always changes exactly 3 output
    # bits — the bit and its two cyclic neighbors. Security comes from FSCX_REVOLVE
    # iteration, not single-step diffusion.
    # Frobenius over GF(2): (1+t+t^-1)^(2^k) = 1+t^(2^k)+t^(-2^k), so power-of-2
    # step counts (like i_val = size//4) also give 3-bit diffusion.
    print("[2] FSCX single-step linear diffusion (expected: exactly 3 bits per flip)")
    for size in SIZES:
        total = 0.0
        gmin = size + 1
        gmax = -1
        for _ in range(1000):
            a = BitArray.random(size)
            b = BitArray.random(size)
            base = fscx(a, b)
            for bit in range(size):
                ap = a.flip_bit(bit)
                hd = (fscx(ap, b) ^ base).popcount()
                total += hd
                if hd < gmin:
                    gmin = hd
                if hd > gmax:
                    gmax = hd
        mean = total / (1000.0 * size)
        status = "PASS" if 2.9 <= mean <= 3.1 else "FAIL"
        print(f"    bits={size:3d}  mean={mean:.2f} (expected 3/{size})  min={gmin}  max={gmax}  [{status}]")
    print()


def test_orbit_period():
    print("[3] Orbit period: FSCX_REVOLVE cycles back to A")
    for size in SIZES:
        cnt_p  = 0
        cnt_hp = 0
        other  = 0
        cap = 2 * size
        for _ in range(100):
            a = BitArray.random(size)
            b = BitArray.random(size)
            cur = fscx(a, b)
            period = 1
            while cur != a and period < cap:
                cur = fscx(cur, b)
                period += 1
            if period == size:
                cnt_p += 1
            elif period == size // 2:
                cnt_hp += 1
            else:
                other += 1
        status = "PASS" if other == 0 else "FAIL"
        print(f"    bits={size:3d}  period={size}: {cnt_p:3d}  period={size//2}: {cnt_hp:3d}  other: {other}  [{status}]")
    print()


def test_bit_frequency():
    print("[4] Bit-frequency bias: 100000 FSCX outputs per size")
    N = 100000
    for size in SIZES:
        counts = [0] * size
        for _ in range(N):
            a = BitArray.random(size)
            b = BitArray.random(size)
            out = fscx(a, b)
            val = out.uint
            for bit in range(size):
                if (val >> bit) & 1:
                    counts[bit] += 1
        pcts = [c / N * 100.0 for c in counts]
        mn = min(pcts)
        mx = max(pcts)
        mean = sum(pcts) / size
        status = "PASS" if mn > 47.0 and mx < 53.0 else "FAIL"
        print(f"    bits={size:3d}  min={mn:.2f}%  max={mx:.2f}%  mean={mean:.2f}%  [{status}]")
    print()


def test_key_sensitivity():
    # sk = FSCX_REVOLVE_N(C2, B, hn, r) ^ A
    # Flipping bit k of A changes sk by exactly 1 bit via the direct XOR term.
    # The nonce change propagates L^i(e_k) into hn = C^C2; algebraically
    # S_r * L^i(e_k) cancels to zero, leaving only the 1-bit XOR contribution.
    # This is a structural property of the HKEX XOR construction.
    print("[5] HKEX session key XOR construction (expected: exactly 1-bit direct sensitivity)")
    for size in SIZES:
        total = 0.0
        iv = i_val(size)
        rv = r_val(size)
        for _ in range(10000):
            a  = BitArray.random(size)
            b  = BitArray.random(size)
            a2 = BitArray.random(size)
            b2 = BitArray.random(size)
            c  = fscx_revolve(a, b, iv)
            c2 = fscx_revolve(a2, b2, iv)
            hn = c ^ c2
            key1 = fscx_revolve_n(c2, b, hn, rv) ^ a
            af   = a.flip_bit(0)
            key2 = fscx_revolve_n(c2, b, hn, rv) ^ af
            total += (key1 ^ key2).popcount()
        mean = total / 10000.0
        status = "PASS" if 0.9 <= mean <= 1.1 else "FAIL"
        print(f"    bits={size:3d}  mean Hamming={mean:.2f} (expected 1/{size})  [{status}]")
    print()


def test_avalanche_revolve_n():
    # FSCX_REVOLVE_N nonce-injection avalanche.
    # Flipping 1 bit of the nonce N while keeping A and B constant propagates
    # through all remaining revolve steps.  The change in the output equals
    # T_n(e_k) where T_n = I + L + ... + L^(n-1) and e_k is the unit vector
    # at bit position k.  Unlike the 3-bit diffusion of single-step FSCX or
    # the purely linear FSCX_REVOLVE, T_n accumulates contributions from every
    # step.  HD = popcount(T_n(e_0)) is deterministic (independent of A and B),
    # so min == max == mean.  For n = size/4 (i_val), HD = size/4 exactly.
    print("[6] FSCX_REVOLVE_N nonce-avalanche: flip 1 nonce bit, measure output diffusion")
    for size in SIZES:
        iv = i_val(size)
        expected = size // 4
        total = 0.0
        gmin = size + 1
        gmax = -1
        for _ in range(1000):
            a = BitArray.random(size)
            b = BitArray.random(size)
            n = BitArray.random(size)
            base    = fscx_revolve_n(a, b, n,           iv)
            nf      = n.flip_bit(0)
            flipped = fscx_revolve_n(a, b, nf,          iv)
            hd = (base ^ flipped).popcount()
            total += hd
            if hd < gmin:
                gmin = hd
            if hd > gmax:
                gmax = hd
        mean = total / 1000.0
        status = "PASS" if mean >= expected else "FAIL"
        print(f"    bits={size:3d}  mean HD={mean:.1f} (expected >={expected})  min={gmin}  max={gmax}  [{status}]")
    print()


# ---------------------------------------------------------------------------
# Performance benchmarks
# ---------------------------------------------------------------------------

def _bench(label: str, fn):
    """Run fn in batches of 100 until ~TARGET_SEC elapsed. Return (ops, elapsed)."""
    # warm up
    for _ in range(10):
        fn()
    t0 = time.perf_counter()
    ops = 0
    while True:
        for _ in range(100):
            fn()
        ops += 100
        elapsed = time.perf_counter() - t0
        if elapsed >= TARGET_SEC:
            break
    rate = ops / elapsed
    if rate >= 1e6:
        rate_str = f"{rate/1e6:.2f} M ops/sec"
    elif rate >= 1e3:
        rate_str = f"{rate/1e3:.2f} K ops/sec"
    else:
        rate_str = f"{rate:.2f} ops/sec"
    print(f"    {label:<40s}: {rate_str}  ({ops} ops in {elapsed:.2f}s)")


def bench_fscx():
    print("[7] FSCX throughput")
    for size in SIZES:
        a = BitArray.random(size)
        b = BitArray.random(size)
        def fn():
            nonlocal a
            a = fscx(a, b)
        _bench(f"bits={size:3d}", fn)
    print()


def bench_fscx_revolve_n():
    print("[8] FSCX_REVOLVE_N throughput")
    for size in SIZES:
        for steps, label in [(i_val(size), f"i({i_val(size)})"), (r_val(size), f"r({r_val(size)})")]:
            a = BitArray.random(size)
            b = BitArray.random(size)
            n = BitArray.random(size)
            def fn(a=a, b=b, n=n, steps=steps):
                return fscx_revolve_n(a, b, n, steps)
            _bench(f"bits={size:3d}  steps={label}", fn)
    print()


def bench_hkex_handshake():
    print("[9] HKEX full handshake")
    for size in SIZES:
        iv = i_val(size)
        rv = r_val(size)
        def fn():
            a  = BitArray.random(size)
            b  = BitArray.random(size)
            a2 = BitArray.random(size)
            b2 = BitArray.random(size)
            c  = fscx_revolve(a, b, iv)
            c2 = fscx_revolve(a2, b2, iv)
            hn = c ^ c2
            _keyA = fscx_revolve_n(c2, b,  hn, rv) ^ a
            _keyB = fscx_revolve_n(c,  b2, hn, rv) ^ a2
        _bench(f"bits={size:3d}", fn)
    print()


def bench_hske_roundtrip():
    print("[10] HSKE round-trip: encrypt+decrypt")
    for size in SIZES:
        iv = i_val(size)
        rv = r_val(size)
        sink = BitArray(size, 0)
        def fn():
            nonlocal sink
            pt  = BitArray.random(size)
            key = BitArray.random(size)
            enc = fscx_revolve_n(pt,  key, key, iv)
            dec = fscx_revolve_n(enc, key, key, rv)
            sink ^= dec ^ pt
        _bench(f"bits={size:3d}", fn)
    print()


# ---------------------------------------------------------------------------
# Entry point
# ---------------------------------------------------------------------------

if __name__ == '__main__':
    print("=== Herradura KEx \u2014 Security & Performance Tests (Python) ===\n")

    print("--- Security Assumption Tests ---\n")
    test_noncommutativity()
    test_avalanche()
    test_orbit_period()
    test_bit_frequency()
    test_key_sensitivity()

    test_avalanche_revolve_n()

    print("--- Performance Benchmarks ---\n")
    bench_fscx()
    bench_fscx_revolve_n()
    bench_hkex_handshake()
    bench_hske_roundtrip()

File: Herradura_tests.go
/*  Herradura KEx -- Security & Performance Tests (Go)

    Copyright (C) 2024-2026 Omar Alejandro Herrera Reyna

    This program is free software: you can redistribute it and/or modify
    it under the terms of the MIT License or the GNU General Public License
    as published by the Free Software Foundation, either version 3 of the License,
    or (at your option) any later version.

    Under the terms of the GNU General Public License, please also consider that:

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

package main

import (
	"crypto/rand"
	"fmt"
	"log"
	"math/big"
	"time"
)

// ---------------------------------------------------------------------------
// BitArray (self-contained)
// ---------------------------------------------------------------------------

type BitArray struct {
	val  big.Int
	size int
}

func bitArrayMask(size int) *big.Int {
	mask := new(big.Int).Lsh(big.NewInt(1), uint(size))
	return mask.Sub(mask, big.NewInt(1))
}

func NewFromBytes(data []byte, _ int, size int) *BitArray {
	ba := &BitArray{size: size}
	ba.val.SetBytes(data[:size/8])
	return ba
}

func New_rand_bitarray(bitlength int) *BitArray {
	buf := make([]byte, bitlength/8)
	_, err := rand.Read(buf)
	if err != nil {
		log.Fatalf("ERROR while generating random string: %s", err)
	}
	return NewFromBytes(buf, 0, bitlength)
}

func randBA(size int) *BitArray {
	return New_rand_bitarray(size)
}

func (ba *BitArray) Xor(other *BitArray) *BitArray {
	result := &BitArray{size: ba.size}
	result.val.Xor(&ba.val, &other.val)
	return result
}

func (ba *BitArray) RotateLeft(n int) *BitArray {
	size := ba.size
	n = ((n % size) + size) % size
	result := &BitArray{size: size}
	if n == 0 {
		result.val.Set(&ba.val)
		return result
	}
	left := new(big.Int).Lsh(&ba.val, uint(n))
	right := new(big.Int).Rsh(&ba.val, uint(size-n))
	result.val.Or(left, right)
	result.val.And(&result.val, bitArrayMask(size))
	return result
}

func (ba *BitArray) Equal(other *BitArray) bool {
	return ba.size == other.size && ba.val.Cmp(&other.val) == 0
}

func (ba *BitArray) Format(f fmt.State, verb rune) {
	hexDigits := ba.size / 4
	s := ba.val.Text(16)
	for i := len(s); i < hexDigits; i++ {
		f.Write([]byte{'0'})
	}
	fmt.Fprint(f, s)
}

// Popcount counts set bits by iterating over bytes.
func (ba *BitArray) Popcount() int {
	cnt := 0
	for _, b := range ba.val.Bytes() {
		v := b
		for v != 0 {
			cnt += int(v & 1)
			v >>= 1
		}
	}
	return cnt
}

// FlipBit returns a new BitArray with bit pos (0=LSB) toggled.
func (ba *BitArray) FlipBit(pos int) *BitArray {
	result := &BitArray{size: ba.size}
	result.val.Set(&ba.val)
	cur := result.val.Bit(pos)
	if cur == 0 {
		result.val.SetBit(&result.val, pos, 1)
	} else {
		result.val.SetBit(&result.val, pos, 0)
	}
	return result
}

// ---------------------------------------------------------------------------
// FSCX functions
// ---------------------------------------------------------------------------

func Fscx(ba, bb *BitArray) *BitArray {
	result := ba.Xor(bb)
	ba = ba.RotateLeft(1)
	bb = bb.RotateLeft(1)
	result = result.Xor(ba).Xor(bb)
	ba = ba.RotateLeft(-2)
	bb = bb.RotateLeft(-2)
	result = result.Xor(ba).Xor(bb)
	return result
}

func FscxRevolve(ba, bb *BitArray, steps int) *BitArray {
	result := ba
	for i := 0; i < steps; i++ {
		result = Fscx(result, bb)
	}
	return result
}

func FscxRevolveN(ba, bb, nonce *BitArray, steps int) *BitArray {
	result := ba
	for i := 0; i < steps; i++ {
		result = Fscx(result, bb).Xor(nonce)
	}
	return result
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

func iVal(size int) int { return size / 4 }
func rVal(size int) int { return size * 3 / 4 }

var sizes = []int{64, 128, 256}

// bench warms up with 10 calls, then runs batches of 100 until 1 second elapsed.
func bench(label string, fn func()) (ops int, elapsed time.Duration) {
	for i := 0; i < 10; i++ {
		fn()
	}
	start := time.Now()
	for {
		for i := 0; i < 100; i++ {
			fn()
		}
		ops += 100
		elapsed = time.Since(start)
		if elapsed >= time.Second {
			break
		}
	}
	return
}

func fmtRate(ops int, elapsed time.Duration) string {
	rate := float64(ops) / elapsed.Seconds()
	if rate >= 1e6 {
		return fmt.Sprintf("%.2f M ops/sec", rate/1e6)
	}
	return fmt.Sprintf("%.2f K ops/sec", rate/1e3)
}

// ---------------------------------------------------------------------------
// Security tests
// ---------------------------------------------------------------------------

func testNonCommutativity() {
	// Fscx(A,B) == Fscx(B,A) always (symmetric formula: A^B^ROL(A)^ROL(B)^ROR(A)^ROR(B)).
	// Asymmetry arises from FscxRevolve, where B is held constant across
	// iterations: FscxRevolve(A,B,n) != FscxRevolve(B,A,n) in general.
	fmt.Println("[1] FscxRevolve non-commutativity: FscxRevolve(A,B,n) != FscxRevolve(B,A,n)")
	for _, size := range sizes {
		iv := iVal(size)
		comm := 0
		for i := 0; i < 10000; i++ {
			a := randBA(size)
			b := randBA(size)
			if FscxRevolve(a, b, iv).Equal(FscxRevolve(b, a, iv)) {
				comm++
			}
		}
		status := "PASS"
		if comm != 0 {
			status = "FAIL"
		}
		fmt.Printf("    bits=%3d  %5d / 10000 commutative  [%s]\n", size, comm, status)
	}
	fmt.Println()
}

func testAvalanche() {
	// Fscx is a linear map over GF(2): output bit i depends only on input bits
	// i-1, i, i+1 (cyclic). Flipping one input bit always changes exactly 3 output
	// bits — the bit and its two cyclic neighbors. Security comes from FscxRevolve
	// iteration, not single-step diffusion.
	// Frobenius over GF(2): (1+t+t^-1)^(2^k) = 1+t^(2^k)+t^(-2^k), so power-of-2
	// step counts (like iVal = size/4) also give exactly 3-bit diffusion.
	fmt.Println("[2] Fscx single-step linear diffusion (expected: exactly 3 bits per flip)")
	for _, size := range sizes {
		total := 0.0
		gmin := size + 1
		gmax := -1
		for trial := 0; trial < 1000; trial++ {
			a := randBA(size)
			b := randBA(size)
			base := Fscx(a, b)
			for bit := 0; bit < size; bit++ {
				ap := a.FlipBit(bit)
				hd := Fscx(ap, b).Xor(base).Popcount()
				total += float64(hd)
				if hd < gmin {
					gmin = hd
				}
				if hd > gmax {
					gmax = hd
				}
			}
		}
		mean := total / (1000.0 * float64(size))
		status := "PASS"
		if mean < 2.9 || mean > 3.1 {
			status = "FAIL"
		}
		fmt.Printf("    bits=%3d  mean=%.2f (expected 3/%d)  min=%d  max=%d  [%s]\n",
			size, mean, size, gmin, gmax, status)
	}
	fmt.Println()
}

func testOrbitPeriod() {
	fmt.Println("[3] Orbit period: FSCX_REVOLVE cycles back to A")
	for _, size := range sizes {
		cntP := 0
		cntHP := 0
		other := 0
		cap := 2 * size
		for trial := 0; trial < 100; trial++ {
			a := randBA(size)
			b := randBA(size)
			cur := Fscx(a, b)
			period := 1
			for !cur.Equal(a) && period < cap {
				cur = Fscx(cur, b)
				period++
			}
			if period == size {
				cntP++
			} else if period == size/2 {
				cntHP++
			} else {
				other++
			}
		}
		status := "PASS"
		if other != 0 {
			status = "FAIL"
		}
		fmt.Printf("    bits=%3d  period=%d: %3d  period=%d: %3d  other: %d  [%s]\n",
			size, size, cntP, size/2, cntHP, other, status)
	}
	fmt.Println()
}

func testBitFrequency() {
	fmt.Println("[4] Bit-frequency bias: 100000 FSCX outputs per size")
	N := 100000
	for _, size := range sizes {
		counts := make([]int, size)
		for trial := 0; trial < N; trial++ {
			a := randBA(size)
			b := randBA(size)
			out := Fscx(a, b)
			for bit := 0; bit < size; bit++ {
				if out.val.Bit(bit) == 1 {
					counts[bit]++
				}
			}
		}
		var mn, mx, mean float64
		mn = 101.0
		mx = -1.0
		for bit := 0; bit < size; bit++ {
			pct := float64(counts[bit]) / float64(N) * 100.0
			mean += pct
			if pct < mn {
				mn = pct
			}
			if pct > mx {
				mx = pct
			}
		}
		mean /= float64(size)
		status := "PASS"
		if mn <= 47.0 || mx >= 53.0 {
			status = "FAIL"
		}
		fmt.Printf("    bits=%3d  min=%.2f%%  max=%.2f%%  mean=%.2f%%  [%s]\n",
			size, mn, mx, mean, status)
	}
	fmt.Println()
}

func testKeySensitivity() {
	// sk = FscxRevolveN(C2, B, hn, r) ^ A
	// Flipping bit k of A changes sk by exactly 1 bit via the direct XOR term.
	// The nonce change propagates L^i(e_k) into hn = C^C2; algebraically
	// S_r * L^i(e_k) cancels to zero, leaving only the 1-bit XOR contribution.
	// This is a structural property of the HKEX XOR construction.
	fmt.Println("[5] HKEX session key XOR construction (expected: exactly 1-bit direct sensitivity)")
	for _, size := range sizes {
		total := 0.0
		iv := iVal(size)
		rv := rVal(size)
		for i := 0; i < 10000; i++ {
			a := randBA(size)
			b := randBA(size)
			a2 := randBA(size)
			b2 := randBA(size)
			c := FscxRevolve(a, b, iv)
			c2 := FscxRevolve(a2, b2, iv)
			hn := c.Xor(c2)
			key1 := FscxRevolveN(c2, b, hn, rv).Xor(a)
			af := a.FlipBit(0)
			key2 := FscxRevolveN(c2, b, hn, rv).Xor(af)
			total += float64(key1.Xor(key2).Popcount())
		}
		mean := total / 10000.0
		status := "PASS"
		if mean < 0.9 || mean > 1.1 {
			status = "FAIL"
		}
		fmt.Printf("    bits=%3d  mean Hamming=%.2f (expected 1/%d)  [%s]\n", size, mean, size, status)
	}
	fmt.Println()
}

// ---------------------------------------------------------------------------
// Performance benchmarks
// ---------------------------------------------------------------------------

func benchFscx() {
	fmt.Println("[6] FSCX throughput")
	for _, size := range sizes {
		a := randBA(size)
		b := randBA(size)
		ops, elapsed := bench(fmt.Sprintf("bits=%3d", size), func() {
			a = Fscx(a, b)
		})
		fmt.Printf("    bits=%3d                          : %s  (%d ops in %.2fs)\n",
			size, fmtRate(ops, elapsed), ops, elapsed.Seconds())
	}
	fmt.Println()
}

func benchFscxRevolve() {
	fmt.Println("[7] FSCX_REVOLVE throughput")
	for _, size := range sizes {
		for _, stepsLabel := range []struct {
			steps int
			label string
		}{
			{iVal(size), fmt.Sprintf("i(%d)", iVal(size))},
			{rVal(size), fmt.Sprintf("r(%d)", rVal(size))},
		} {
			a := randBA(size)
			b := randBA(size)
			steps := stepsLabel.steps
			lbl := stepsLabel.label
			ops, elapsed := bench("", func() {
				FscxRevolve(a, b, steps)
			})
			fmt.Printf("    bits=%3d  steps=%-12s        : %s  (%d ops in %.2fs)\n",
				size, lbl, fmtRate(ops, elapsed), ops, elapsed.Seconds())
		}
	}
	fmt.Println()
}

func benchHkexHandshake() {
	fmt.Println("[8] HKEX full handshake")
	for _, size := range sizes {
		iv := iVal(size)
		rv := rVal(size)
		ops, elapsed := bench("", func() {
			a := randBA(size)
			b := randBA(size)
			a2 := randBA(size)
			b2 := randBA(size)
			c := FscxRevolve(a, b, iv)
			c2 := FscxRevolve(a2, b2, iv)
			hn := c.Xor(c2)
			_ = FscxRevolveN(c2, b, hn, rv).Xor(a)
			_ = FscxRevolveN(c, b2, hn, rv).Xor(a2)
		})
		fmt.Printf("    bits=%3d                          : %s  (%d ops in %.2fs)\n",
			size, fmtRate(ops, elapsed), ops, elapsed.Seconds())
	}
	fmt.Println()
}

func benchHskeRoundTrip() {
	fmt.Println("[9] HSKE round-trip: encrypt+decrypt")
	for _, size := range sizes {
		iv := iVal(size)
		rv := rVal(size)
		sink := randBA(size)
		ops, elapsed := bench("", func() {
			pt := randBA(size)
			key := randBA(size)
			enc := FscxRevolveN(pt, key, key, iv)
			dec := FscxRevolveN(enc, key, key, rv)
			sink = sink.Xor(dec.Xor(pt))
		})
		fmt.Printf("    bits=%3d                          : %s  (%d ops in %.2fs)\n",
			size, fmtRate(ops, elapsed), ops, elapsed.Seconds())
	}
	fmt.Println()
}

// ---------------------------------------------------------------------------
// main
// ---------------------------------------------------------------------------

func main() {
	fmt.Println("=== Herradura KEx \u2014 Security & Performance Tests (Go) ===\n")

	fmt.Println("--- Security Assumption Tests ---\n")
	testNonCommutativity()
	testAvalanche()
	testOrbitPeriod()
	testBitFrequency()
	testKeySensitivity()

	fmt.Println("--- Performance Benchmarks ---\n")
	benchFscx()
	benchFscxRevolve()
	benchHkexHandshake()
	benchHskeRoundTrip()
}
